sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/query/DocumentQueryExecutionContextFactory.java

```diff
@@ -60,37 +66,82 @@
             collectionObs = resolveCollection(client, query, resourceTypeEnum, resourceLink).flux();
         }
 
-        // We create a ProxyDocumentQueryExecutionContext that will be initialized with DefaultDocumentQueryExecutionContext
-        // which will be used to send the query to GATEWAY and on getting 400(bad request) with 1004(cross parition query not servable), we initialize it with
-        // PipelinedDocumentQueryExecutionContext by providing the partition query execution info that's needed(which we get from the exception returned from GATEWAY).
-
-        Flux<ProxyDocumentQueryExecutionContext<T>> proxyQueryExecutionContext =
-                collectionObs.flatMap(collection -> {
-                    if (feedOptions != null && feedOptions.partitionKey() != null && feedOptions.partitionKey().equals(PartitionKey.None)) {
-                        feedOptions.partitionKey(BridgeInternal.getPartitionKey(BridgeInternal.getNonePartitionKey(collection.getPartitionKey())));
-                    }
-                    return ProxyDocumentQueryExecutionContext.createAsync(
-                            client,
-                            resourceTypeEnum,
-                            resourceType,
-                            query,
-                            feedOptions,
-                            resourceLink,
-                            collection,
-                            isContinuationExpected,
-                            correlatedActivityId);
-                    }).switchIfEmpty(ProxyDocumentQueryExecutionContext.createAsync(
-                            client,
-                            resourceTypeEnum,
-                            resourceType,
-                            query,
-                            feedOptions,
-                            resourceLink,
-                            null,
-                            isContinuationExpected,
-                            correlatedActivityId));
-
-        return proxyQueryExecutionContext;
+        DefaultDocumentQueryExecutionContext<T> queryExecutionContext = new DefaultDocumentQueryExecutionContext<T>(
+            client,
+            resourceTypeEnum,
+            resourceType,
+            query,
+            feedOptions,
+            resourceLink,
+            correlatedActivityId,
+            isContinuationExpected);
+
+        if (ResourceType.Document != resourceTypeEnum) {
+            return Flux.just(queryExecutionContext);
+        }
+
+        Mono<PartitionedQueryExecutionInfo> queryExecutionInfoMono =
+            com.azure.data.cosmos.internal.query.QueryPlanRetriever.getQueryPlanThroughGatewayAsync(client, query, resourceLink);
+
+        return collectionObs.single().flatMap(collection ->
```

Comment:
Earlier there was a switchIfEmpty() block because of collectionObs is initialized with Flux.empty()
so in case it is not a child collection, collectionObs will still be empty, and you will never enter in this flatMap. 
We should fix it by adding a switchIfEmpty() block here, and doing the common work in that block.

Reply:
Everything non document goes through the gateway which is the DefaultDocumentQueryExecutionContext. Only document goes through the queryplan and the parallel pipeline